src/video/cocoa/SDL_cocoawindow.m

```diff
@@ -2230,33 +2230,24 @@ - (BOOL)acceptsFirstMouse:(NSEvent *)theEvent
 Cocoa_GetWindowDisplayIndex(_THIS, SDL_Window * window)
 { @autoreleasepool
 {
-    NSRect displayframe;
-    SDL_Point display_center;
-    SDL_Rect sdl_display_rect;
     SDL_WindowData *data = (__bridge SDL_WindowData *) window->driverdata;
 
     /* Not recognized via CHECK_WINDOW_MAGIC */
     if (data == NULL){
         return 0;
     }
 
-    /*
-     Considering that we already have the display coordinates in which the window is placed (described via displayframe)
-     instead of checking in which display the window is placed, we should check which SDL display matches the display described
-     via displayframe.
-    */
-    displayframe = data.nswindow.screen.frame;
-    
-    display_center.x = displayframe.origin.x + displayframe.size.width / 2;
-    display_center.y = displayframe.origin.y + displayframe.size.height / 2;
-    
-    for (int i = 0; i < SDL_GetNumVideoDisplays(); i++){
-        SDL_GetDisplayBounds(i, &sdl_display_rect);
-        if (SDL_EnclosePoints(&display_center, 1, &sdl_display_rect, NULL)) {
-            return i;
-        }
+    NSArray *screens = [NSScreen screens];
+
+    int index = 0;
+    for (NSScreen *screen in screens) {
+        if (screen == data.nswindow.screen)
+            return index;
+
+        index++;
```

Comment:
I wonder if it's ever possible for the `[NSScreen screens]` array order to mismatch (or have a different count from) the higher level SDL video display APIs. I haven't looked at the relevant code to see if that could be possible (for example in display disconnect/reconnect situations), but maybe there's some way to iterate through the SDL display indices while still using the `NSScreen *` associated with that SDL display index?

Reply:
I had similar concerns but had trouble following the sdl code pathway for display indexing so a second check would be appreciated.

I tried my best to confirm there would not be a mismatch (through manual testing).

Reply:
I believe this approach should work pretty robustly, if you don't mind testing it. I've only tested with a single monitor:
```patch
diff --git a/src/video/cocoa/SDL_cocoawindow.m b/src/video/cocoa/SDL_cocoawindow.m
index dca560ba9e95ba1b9f5bf4b576d674358ed48fcb..157353cf0da75d0031a6fa5afba0b811ce1cf43e 100644
--- a/src/video/cocoa/SDL_cocoawindow.m
+++ b/src/video/cocoa/SDL_cocoawindow.m
@@ -2230,34 +2230,37 @@ int
 Cocoa_GetWindowDisplayIndex(_THIS, SDL_Window * window)
 { @autoreleasepool
 {
-    NSRect displayframe;
-    SDL_Point display_center;
-    SDL_Rect sdl_display_rect;
+    NSScreen *screen;
     SDL_WindowData *data = (__bridge SDL_WindowData *) window->driverdata;
 
     /* Not recognized via CHECK_WINDOW_MAGIC */
-    if (data == NULL){
+    if (data == nil) {
         return 0;
     }
 
-    /*
-     Considering that we already have the display coordinates in which the window is placed (described via displayframe)
-     instead of checking in which display the window is placed, we should check which SDL display matches the display described
-     via displayframe.
-    */
-    displayframe = data.nswindow.screen.frame;
-    
-    display_center.x = displayframe.origin.x + displayframe.size.width / 2;
-    display_center.y = displayframe.origin.y + displayframe.size.height / 2;
-    
-    for (int i = 0; i < SDL_GetNumVideoDisplays(); i++){
-        SDL_GetDisplayBounds(i, &sdl_display_rect);
-        if (SDL_EnclosePoints(&display_center, 1, &sdl_display_rect, NULL)) {
-            return i;
+    /* NSWindow.screen may be nil when the window is off-screen. */
+    screen = data.nswindow.screen;
+
+    if (screen != nil) {
+        CGDirectDisplayID displayid;
+        int i;
+
+        /* https://developer.apple.com/documentation/appkit/nsscreen/1388360-devicedescription?language=objc */
+        displayid = [[screen.deviceDescription objectForKey:@"NSScreenNumber"] unsignedIntValue];
+
+        for (i = 0; i < _this->num_displays; i++) {
+            SDL_DisplayData *displaydata = (SDL_DisplayData *)_this->displays[i].driverdata;
+            if (displaydata != NULL && displaydata->display == displayid) {
+                return i;
+            }
         }
     }
-    SDL_SetError("Couldn't find the display where the window is attached to.");
-    return -1;
+
+    /* Other code may expect SDL_GetWindowDisplayIndex to always return a valid
+     * index for a window. The higher level GetWindowDisplayIndex code will fall
+     * back to a generic position-based query if the backend implementation
+     * fails. */
+    return SDL_SetError("Couldn't find the display where the window is located.");
 }}
 
 int
diff --git a/src/video/SDL_video.c b/src/video/SDL_video.c
index 1e520b24b68b3b743effd3ea98b8ffcfd399d322..144326f5434a5bfc7e14bd37450b5439b72cbe26 100644
--- a/src/video/SDL_video.c
+++ b/src/video/SDL_video.c
@@ -1071,11 +1071,20 @@ SDL_GetDisplay(int displayIndex)
 int
 SDL_GetWindowDisplayIndex(SDL_Window * window)
 {
+    int displayIndex = -1;
+
     CHECK_WINDOW_MAGIC(window, -1);
     if (_this->GetWindowDisplayIndex) {
-        return _this->GetWindowDisplayIndex(_this, window);
+        displayIndex = _this->GetWindowDisplayIndex(_this, window);
+    }
+
+    /* A backend implementation may fail to get a display index for the window
+     * (for example if the window is off-screen), but other code may expect it
+     * to succeed in that situation, so we fall back to a generic position-
+     * based implementation in that case. */
+    if (displayIndex >= 0) {
+        return displayIndex;
     } else {
-        int displayIndex;
         int i, dist;
         int closest = -1;
         int closest_dist = 0x7FFFFFFF;

```

Reply:
@slime73 It seems to work well in my test scenarios. If that's a preferred direction, let me know if you'd like me to commit that here or if you'd prefer to PR from scratch.

Reply:
Committing it to this PR is fine with me